### PR TITLE
Updates and clarification for the results section

### PR DESCRIPTION
**Pull request by gkiar**
Hi @HamidHeyde ,

I've been going through the Results section and figured I'd open a PR right away so we can track questions that emerge and clarifications that are needed while I go along. I'll update this comment with questions as I come up with them, and will ping you either here or on Slack if/when I add new ones. Thanks for bearing with me through these!

### Questions:

- [x] How are we defining the "best" models? Based on the tables of performance, no two of our models are performing the exact same as each other, or have the same "distance" value, so I am unclear on a) how we are determining the "best" one(s), and b) why we have more than one when they aren't equal.
> Model clusters are chosen to avoid overfitting. They are the ones which minimize distance. New points:
> - [x] We need to explicitly acknowledge groupings/clusters of results and how they are defined.
> - [x] Sensitivity (and other scores) really range from 0 - 1, but we're representing them as though they're 0 - 100 (yet still using MCC on a -1 to 1 scale). This *needs* to be made consistent across features and I mention it again below.

- [x] Related to the above, in the tables, how is "distance" defined? Presumably it is the distance from reference performance, but it is unclear what feature of the reference performance is being used for that.
> Euclidean Distance. New points:
> - [x]  The data *NEED* to be normalized, since MCC is on a much smaller scale than the other scores it currently has negligible impact on results.
> - [x] As @glatard mentioned, there is certainly some redundancy in the performance values so it would be worth doing something similar to a dimensional analysis where we look at the minimum set of values that include unique information/don't just scale with one another, and compute the distance between those.

- [x] Can you please provide the gamma and cost values associated with each row of the tables? I understand that these should **not** be a part of *these* tables, as in the right half (variable gamma & cost) this would be 7 or 8 values per row, but if you **could send me a CSV with a row per model and a column per a) the defining features of that model, and b) the gamma and cost values**, then I can find a way to include them. We need to report these values so that we can comment on their importance in model parameterization, as we noted it was something the original authors did not report and these values comprise a space we are exploring in our evaluation.
> Spreadsheet [linked here](https://docs.google.com/spreadsheets/d/1vojVBWc4DU_2E4mKI1PbIsiGs3PwUvLzrIXsCXYbq0I/edit?usp=sharing).

- [ ] You mention that the balanced average (or, referred to it by its old name, "threshold-based aggregation") outperforms the other methods on both the shuffled and balanced datasets. However, in a balanced dataset, the threshold-based/balanced should be equivalent to the class-based/un-weighted average, as the classes/thresholds are already balanced. Can you please clarify what's happening here? It is important to verify this is the case to either a) update our understanding and explanation of what these different aggregation methods are actually doing, or b) root out possible bugs in our analysis.
> It is still unclear what is happening here. Based on previous discussions, the balanced-average appeared to threshold classes based on their membership size, whereas the description below suggests that *all* classes are given the same non-zero threshold. The justification for using the same threshold for each class (importantly, in an unbalanced setting) is not clear to me at present. As I said, based on my previous understanding of what balanced-average meant, the results *are* equivalent in a balanced and unweighted context, but now I'm not particularly sure. I think this needs a real whiteboard discussion with all three of us and concrete numbers that can be used to explain what is happening.

- [x]  Which model was actually used when extending our analysis to use more features than just ACC? This is related to the first comment about where we confusingly have a few "best" models, thus it is ambiguous what we are using here.
> We chose the one which minimizes distance. New point:
> - [ ] If we're clustering the "best models" to avoid over fitting, and then just picking a single "best model" anyways for the downstream results, we're kind of defeating our own purpose. I think that we *need* to do everything downstream with all of the models in our best cluster if we're serious about not overfitting.

- [x] How come, when applying the best model to more features, we never fit with all 5 features at once? Didn't the original authors do this? It looks like we're using either pairs or triples, but never a quad or all five. The justification for this is currently very unclear.
- [x] Did the original paper ever report on performance using all features? It doesn't show up anywhere in the current Table 4 (which the above bullet refers to), which kind of begs the question as to why we are evaluating all of these intermediate combinations of variables if all we have/are comparing to is a single performance value here.
> The original paper never did this. This would be a great place to explore methodological flexibility but is presently beyond the scope - I think we should do it while the paper is under review.

### Eventual ToDos (before merge):
- [x] We should reorganize the tables and make them more concise/clear. I will propose a layout for this before submitting my PR. **Could you also send me a CSV with all of this performance data**?

- [x] We should re-plot our data on a single set of axes and encode the number of classes as a feature of the plot. I will also develop a prototype for this before submitting the PR.

- [ ] As mentioned in the last review of our methods section, we need to evaluate all of our models on the (originally referred to as-) independent set, and report on these results and potentially have another figure showing the sens/spec for these, as well.

### CODE REVIEW

**Comment by HamidHeyde** [comment]
> How are we defining the "best" models? Based on the tables of performance, no two of our models are performing the exact same as each other, or have the same "distance" value, so I am unclear on a) how we are determining the "best" one(s), and b) why we have more than one when they aren't equal.

Regarding the best Model,  The best models are the closest ones (Distance), Since there are more than one Metrics, and not all the 4 values are comparable to each other on different models, what @glatard proposed was to use a distance which is the distance from the our results from the published original ones `(Original results on all Tables)` which can be used through all different context of the paper. (axis of flexibility and which one provides the closest results which is the base to the idea of the model we decided to move forward with for  the rest of the experiments, and how Scikit-learn is comparable with SVMLight through the same context.)

But when it comes to which model performs better overall, the one with 8-class (threshold-based which was the closest one), produces higher values when compared to the same one with 7-classes in the dataset.
So, i think, we can say, (1) i.e. these models (regardless of how many classes present in the dataset) provides the closest results, but if we do the experiment with 8 class, the results improves a bit. and (2) then we can say, since this seems to be right model, then if we want to get to the same results with scikit-learn then we should do it in this settings and here are the results. and we can finally end it with, if we do these or report on these on the paper, we won't have that much flexibility.

I agree that it's a bit confusing when looking at all these results we put together in these tables, but, then looking at those tables from this perspective, then we can find best models. The red ones on Table 2&3 on Threshold based models (svmlight, scikit), provide closest distance, and all close to each other. So, if we put 8-classes in the dataset and use SVMLight then on threshold-based aggregation (closest models) there are 2 models with 0.10 distance from the original results and 2 with the 0.09 distance. the same setting with scikit-learn there are 3 models with 0.08 distance. same thing with 7-classes, there are 2 models with 0.09 distance when classified with SVMLight and 3-models with 0.07 distance when classified with scikit-Learn.

and all these could be chosen as best (closest)

**Comment by HamidHeyde** [comment]
> Related to the above, in the tables, how is "distance" defined? Presumably it is the distance from reference performance, but it is unclear what feature of the reference performance is being used for that.

Regrading the Distance (issue no.2): it's the distance in between the published original 4 metrics values (for AAC Feature), and the models we experimented on.(i.e. accuracy, sensitivity, specificity, mcc of the published one (73.74- 74.65- 73.22 -0.46)  <= and => threshold-based with 7-classes (line 7 Table 2) 73.84- 75.25 -73.60 -0.36 which Equals to => 0.09)

**Comment by HamidHeyde** [comment]
> Can you please provide the gamma and cost values associated with each row of the tables? I understand that these should not be a part of these tables, as in the right half (variable gamma & cost) this would be 7 or 8 values per row, but if you could send me a CSV with a row per model and a column per a) the defining features of that model, and b) the gamma and cost values, then I can find a way to include them. We need to report these values so that we can comment on their importance in model parameterization, as we noted it was something the original authors did not report and these values comprise a space we are exploring in our evaluation.


Regrading the CSV for Models, I created a spreadsheet (https://docs.google.com/spreadsheets/d/1vojVBWc4DU_2E4mKI1PbIsiGs3PwUvLzrIXsCXYbq0I/edit?usp=sharing) and i'll add all the results and models to this one. Hopefully, by the end of the week, i'll be done with all the new experiments. Since there are lot of models, and some needed modifications it took longer than expected. So, ill add the new ones to the current ones on the paper, the Gamma and Cost for all of them to that spreadsheet i shared the link above by the end of the week.

**Comment by HamidHeyde** [comment]
> How come, when applying the best model to more features, we never fit with all 5 features at once? Didn't the original authors do this? It looks like we're using either pairs or triples, but never a quad or all five. 

The authors in the original paper haven't done it using all 5 features. (original paper => introduction, Table 1)

> The justification for this is currently very unclear.

I think, considering the number of sequences in the dataset and the imbalance nature of the set, adding more features means adding more dimension, which doesn’t seem to improve the results considerably compared to what they have provided on their Table1 of the original paper. Also, by trying pairs or triples, they came to a conclusion that a specific pair outperforms the others. That could be another reason why they didn’t go any further with adding features. (original paper => introduction, Table 1)

We can email the authors and ask for more info over the matter, but whatever the justification, I don't think that would affect what we are doing, and what we are trying to address in this paper in any way.

**Comment by HamidHeyde** [comment]
> Did the original paper ever report on performance using all features? It doesn't show up anywhere in the current Table 4 (which the above bullet refers to), which kind of begs the question as to why we are evaluating all of these intermediate combinations of variables if all we have/are comparing to is a single performance value here.

I think the answer to this question is the same as the one [above](https://github.com/big-data-lab-team/reproducibility-bioinfo/pull/37#issuecomment-649558534). And still, I can’t see how this is related to / affects what we are doing in this paper.

If you are wondering why we are trying all these models, then because, we need to do so to address the axis of flexibility available in the paper. But since for each model (Table 4 of our paper, and Table1 of original paper) we can’t test and go through all these scenarios to find the closest one, then,  for one feature (AAC) we tried all those different scenarios, and then we chose the closest model settings, for the rest of them.

**Comment by HamidHeyde** [comment]
> Which model was actually used when extending our analysis to use more features than just ACC?

The closest results to the original paper (regardless of the number of the classes in the dataset) is achieved when aggregated using threshold-based technique on balanced versions of the dataset using SVMLight software (the one being used in the original study).


When we have 7-classes in the dataset, the accuracy, sensitivity and specificity values are close to the original one but MCC is lower. When we have 8-classes in the dataset, the accuracy, sensitivity and specificity values are higher compared to the original one but MCC is close.

Since in the Table 1 of the original paper it says “The average sensitivity, specificity, accuracy, and MCC for all seven substrate-specific transporter classes for different models” then we chose to go with this one for the rest of the models.


> This is related to the first comment about where we confusingly have a few "best" models, thus it is ambiguous what we are using here.

So, I guess, I have explained why we have few best models and how we should look at them in the context of this paper in this [comment](https://github.com/big-data-lab-team/reproducibility-bioinfo/pull/37#issuecomment-648989018). But let me know if you needed any further explanation over the matter.

**Comment by HamidHeyde** [comment]
> You mention that the balanced average (or, referred to it by its old name, "threshold-based aggregation") outperforms the other methods on both the shuffled and balanced datasets. 

it provides the closest results to the original published ones, through that setting. We want to say, if aggregation technique is not not presented, then through an imbalanced setting, this range of results are possible.

> However, in a balanced dataset, the threshold-based/balanced should be equivalent to the 
class-based/un-weighted average, as the classes/thresholds are already balanced. 

Not necessarily.

> Can you please clarify what's happening here? It is important to verify this is the case to either a) update our understanding and explanation of what these different aggregation methods are actually doing, or b) root out possible bugs in our analysis.

So,  after classification, the binary classifier provides a list of probabilities, that by default the ones ” >0 “ would be considered as => +1 label and the ones “<0” are considered => -1 label. when compared with original labels then for each class we can calculate tp,tn,fp,fn values and from those you can calculate metrics for that class. I labeled these default metrics as => class-based. Which provides an unbalanced sensitivity/specificity value pair.

To balance sensitivity/specificity, then you can define a single threshold value for all the classes. i.e. 0.5. Then for example, the ones ” >0.5 “ would be considered as => +1 label and the ones “<0.5” would be considered as => -1.  Then you calculate tp,tn,fp,fn and then metrics. I labeled this as => threshold-based.

Basically here, you don’t aggregate results, and you take results from each binary classifier into consideration (multi-label classification problem).

But when you aggregate, you are looking for a single prediction out of multiple results, applying a threshold does not affect the results here (multi-class classification problem). I labeled this as => vote-based (based on the aggregation technique)

The balanced, shuffled and down-sampled are the order and number of sequences in the dataset. The dataset can include 7 or 8 classes.

I think we can look at these from this perspective: If not stated, in an imbalanced dataset, each one can produce these results.

**Comment by glatard** [comment]
### About the best model(s)

#### Comments about discussion above
* I believe @HamidHeyde computed the distance as the Euclidean distance in the 4D space made from accuracy, sensitivity, specificity and mcc. 
* There is more than a single "best" model because we should think of this as a cluster of best models, to avoid overfitting.

#### Suggestions
* We should use a different terminology to avoid ambiguity. Maybe "closest" or "most reproducible".
* I don't think it makes much sense to compute this distance in a 4D space, since specificity+sensitivity are redundant with at least accuracy and most likely MCC. @HamidHeyde, is it possible that two models have the same sensitivity, the same specificity, but different MCCs? If not, then we should compute the distances on sensitivity and specificity only, and discuss MCC from the tables. That would also make that discussion possible from just Figure 1.
* We could rank the models with this distance in the table and discuss model clusters from ranks.
* From Figure 1, 7 classes, it is clear that the closest models are 3 SVM light models and 1 SKlearn one. It would be interesting to represent the discriminative parameters of these models on the figure, so that we can clearly visualize what made these models the closest ones.

**Comment by gkiar** [comment]
Hi @HamidHeyde - thanks for the comments. I updated my initial comment with some notes/answers extracted from your comments here. Please re-read the new version of my initial PR comment and respond to the next round of questions/comments when you have a chance.

**Comment by HamidHeyde** [comment]
> If we're going to combine 4 parameters in a distance calculation, these parameters need to be normalized first. In your calculation @HamidHeyde, MCC has negligible impact on the distance because it's in [0,1] while all the other parameters are in [0,100]

@glatard : I calculated the Euclidean distances in a 4D. For each model, the value shows the distance in between the original published Accuracy, Sensitivity, specificity and mcc for AAC feature (73.74, 74.65, 73.22, 0.46) and accuracy, sensitivity, specificity, mcc of each model.

For the distance value, i normalized all the 4 values in the range of [0,1] and then calculated the final distance. But since “Mishra et al.” have presented the values in that format, I did the same. So, we could kind of visually compare in between these models and the original one when needed.


>We should use a different terminology to avoid ambiguity. Maybe "closest" or "most reproducible".

I think the “closest” fits best in this context.

>I don't think it makes much sense to compute this distance in a 4D space,  since specificity+sensitivity are redundant with at least accuracy and most likely MCC. @HamidHeyde, is it possible that two models have the same sensitivity, the same specificity, but different MCCs? If not, then we should compute the distances on sensitivity and specificity only, and discuss MCC from the tables. That would also make that discussion possible from just Figure 1.

I checked the models, and it seems that all the models with the same specificity and sensitivity have the same MCC. I originally computed the distance in 4D to take all the metrics into consideration. I agree we can ignore accuracy as it will always provide a high value regardless of what model we use/develop. But I'm not sure how to take MCC into consideration, if we disclude it from the distance calculation as one of the main problems was the MCC value and the reason we tried the 8-class based models was to somehow increase the MCC value and see if we can get closer to the original MCC value. @glatard please advise on how to do this and how to include MCC into this.

I have already calculated the distance using just the Sensitivity and specificity (as you mentioned) just for the models based on 7 classes using Scikit-learn software (scikit-7 sheet in [this](https://docs.google.com/spreadsheets/d/1vojVBWc4DU_2E4mKI1PbIsiGs3PwUvLzrIXsCXYbq0I/edit?usp=sharing) Google spreadsheet). Please find/check the difference in distance values there. 

>We could rank the models with this distance in the table and discuss model clusters from ranks.
>From Figure 1, 7 classes, it is clear that the closest models are 3 SVM light models and 1 SKlearn one. It would be interesting to represent the discriminative parameters of these models on the figure, so that we can clearly visualize what made these models the closest ones.

I agree. Generally, regardless of the software (scikit or SVMLight), the Threshold-based model results fall in one cluster closer than the others to the original ones. With that in mind, the micro averaged results (on balanced and shuffled sets) are the closest to the original ones.

**Comment by HamidHeyde** [comment]
>Model clusters are chosen to avoid overfitting. They are the ones which minimize distance. New points:  We need to explicitly acknowledge groupings/clusters of results and how they are defined.

I’m not quite sure if i completely understand this point. In figures on the paper, they are grouped based on the sensitivity/specificity results. So, you mean more diagrams/figures? Like another one based on distance from the original one? Or any others?

For overfitting/under-fitting normally we can check that (for each model) on the test/validation set which is available on [this](https://docs.google.com/spreadsheets/d/1vojVBWc4DU_2E4mKI1PbIsiGs3PwUvLzrIXsCXYbq0I/edit?usp=sharing) Google spreadsheet.

>Sensitivity (and other scores) really range from 0 - 1, but we're representing them as though they're 0 - 100 (yet still using MCC on a -1 to 1 scale). This needs to be made consistent across features and I mention it again below.

>The data NEED to be normalized, since MCC is on a much smaller scale than the other scores it currently has negligible impact on results.

For the distance value, i normalized all the 4 values in the range of [0,1] and then calculated the final distance. But since “Mishra et al.” have presented the values in that format, I did the same. So, we could kind of visually compare in between these models and the original one when needed.

>As @glatard mentioned, there is certainly some redundancy in the performance values so it would be worth doing something similar to a dimensional analysis where we look at the minimum set of values that include unique information/don't just scale with one another, and compute the distance between those.

As a data preprocessing step, exploring Feature-selection is relevant here, but it has been mostly used in Genomics cause the dimension of i.e. 20,000D is not strange in i.e. microarray problems. But we didn’t explore it here. (we can add this one too but then we need to discuss which technique and how...)

>You mention that the balanced average (or, referred to it by its old name, "threshold-based aggregation") outperforms the other methods on both the shuffled and balanced datasets. However, in a balanced dataset, the threshold-based/balanced should be equivalent to the class-based/un-weighted average, as the classes/thresholds are already balanced. Can you please clarify what's happening here? It is important to verify this is the case to either a) update our understanding and explanation of what these different aggregation methods are actually doing, or b) root out possible bugs in our analysis.

By outperforming, I mean producing the closest results compared with the original published ones.

On a multi-Label classification problem (when we can have more than one predicted label for a sequence), the software used for classification, produces a probability value for each sequence. Since this is a binary classifier, using that value a label of +1 or -1 will be assigned to that sequence/element. The software uses a default threshold (i.e. for SVMlight the values > 0 are considered as +1 and <0 are considered as -1) and produces the confusion matrix values (tp, fn, tn, fp) from that. Now, you can move this threshold value ( for example, for SVMlight, we can assign +1 label to values > -0.5 and assign -1 Label to values <-0.5). 

So, regardless of the dataset setting (balanced, etc.), when we use the default threshold value used by the software (class-based) that would give us a result. Which is not close to the original published one (also sensitivity and specificity value pair are not in balance). Now the question we are trying to answer is that: how can we balance out  the sensitivity and specificity value pair? what if we move that threshold value for all binary classifiers? Then would it balance out those two? And then how would it affect the other performance metrics? And would it in the end of the day provide closer results or farther results when the sensitivity and specificity values are in balance. This is the Threshold-based model.

You are right, this is an unweighted problem and we did not assign any weight to any class since there was not any clue on this in the paper. Going through that idea is not really possible, cause then we need to know what weight to assign to which class, which is not really possible. So, we tried and got really close to the published result without any weight, so we ignored that idea.


>If we're clustering the "best models" to avoid over-fitting, and then just picking a single "best model" anyways for the downstream results, we're kind of defeating our own purpose. I think that we need to do everything downstream with all of the models in our best cluster if we're serious about not overfitting.

I think these are two different problems and we are not exploring overfitting here in this study:

- First, We want to see in the context of reproducibility, how different “parameters/steps/axis of analysis”  involved in creating a machine-learning model, would affect the final results, if missed or not reported properly in an imbalanced setting. 

(The importance of an “Imbalanced settings” is that some of these “parameters/steps/axis of analysis” does not matter much (if you miss to report on them) in a balanced setting. The down-sampled set models’ results are a good evidence on that subject. )

- And second, eventually based on what we discussed above, we need to use/choose one of the settings and produce results for other features as well. so, we could see if what we have chosen would provide the same thing for the other features in comparison to the original published one for that feature. We are trying to kind of provide proof over what discussed above and say that that is true. 
(for this one, we can either 

- - choose one of our best models and use that for all other features, 
- - or calculate all models for all other features as well, 
- - or calculate all the models for one other feature,
- - or simply ignore it. But then how should we provide proof for our discussion above? This is (from my point of view) kind of proves our point on a real case with available published results.

**Comment by HamidHeyde** [comment]
Here, I’ll go over what we are trying to do in this paper (from my point of view), which I think will clear some confusion points. I agree with @gkiar, i think as for the next step with all the results in hand ([this](https://docs.google.com/spreadsheets/d/1vojVBWc4DU_2E4mKI1PbIsiGs3PwUvLzrIXsCXYbq0I/edit?usp=sharing) Google spreadsheet), we need a zoom meeting to go through the rest, because, i think some concepts are “hard to/couldn’t be” communicated clearly through some few lines of text.

To my Understanding, what we are doing here is that we want to study:

In the context of reproducibility, how different “parameters/steps/axis of analysis”  involved in creating a machine-learning model, would affect the final results, if missed or not reported properly in an imbalanced setting. 

The importance of an “Imbalanced settings” is that some of these “parameters/steps/axis of analysis” does not matter much (if you miss to report on them) in a balanced setting. The down-sampled set models’ results are a good evidence on that subject. 

So, to do this we need a work to base our study on. Something with a reasonable amount of information for such a study. Something relevant. The study i picked for this purpose has following properties:

- It’s in the bioinformatics and proteomics domain.
- The subject is Prediction of Membrane Transport Proteins and Their Substrate Specificities which is still relevant and people are trying to improve the work (Munira from Dr. Butler lab, published a paper in this line of study in 2020).
- In this domain of work (proteomics, sequence-based study), normally the dataset is imbalanced
- One of the popular classification algorithm in this field is SVM

So, according to what i just mentioned above, this type of problem can be something that scientist would face in their studies and we want to imply that since this is common and relevant in this line of work, then,  it would make it more reproducible if you take these things into consideration when reporting on the study. Or else the following range of results are probable, depending on which element you miss to clarify/report in your paper.

The axises in a common model on an imbalance setting are:
- Dataset: normally no scientist would miss on attaching the dataset elements to their report but the strategy for dividing the dataset into train/test sets could be missed which can cause problems (specially on dataset with huge number of elements). We can:
  - Shuffle the whole dataset and then divide them into train and test (causes some class to be present in training set and not present in the test set)
  - I.e. for train/test split of 80%|20%, we take 80% from each class and put them together in the training set and do the same thing for test set with 20% ratio (so we could make sure that the same ration of each class would be present in train/test)
  - If there has been any down-sampling / over-sampling in an effort to balance out the dataset (the oversampling basically is not really possible in genomics and proteomics, because if we do so, then it means we are creating new proteins or genoms which does not exist)

- Generation of features from the dataset
  - Based on the formula available for each feature, then we can generate the numerical value for each sequence. I think since the features are well-known even if they have not been mentioned in detail on the paper, it can be generated by consulting with people in the field of biology. So, the only thing we can discuss in this regard, is, please make sure the details are available (which is a point in my discussion)

- Data pre-processing
  - Exploring Feature-selection is relevant here, but it has been mostly used in Genomics cause the dimension of i.e. 20,000D is not strange in i.e. microarray problems. But we didn’t explore it here. (we can add this one too but then we need to discuss which technique and how...)

- Learning/Training and Classification
  - The first step in to create multiple binary classifiers. There are 2 common strategies in practice:
    - One vs all 
    - One vs one (this one is much more expensive [ time-wise and computation-wise ] compared to the other one)
  - Hyper-parameters of the classification algorithm (in this case Gamma and Cost for RBF kernel)
  - The software (SVMLight vs ScikitLearn’s SVC): Because they apply different default thresholds to classification results and therefore produce different but close probability values

- Aggregation technique
  - Multi-label classification (we can have more than one predicted label for an element)
  - Threshold-based => So, in multi-label problems, each binary classifier, provides a prediction value (probability) for each element. And by applying a default (software-based) threshold, it will produce the confusion matrix (tp, fn, tn, fp) values. We can somehow move that threshold, but we need to move it for all the binary classifiers to the same amount.
  - Multi-class classification. (we can have only one predicted Label for each element)

- Calculation of final results
  - The performance metrics could be averaged using `Micro`or `Macro` technique which could cause a difference in imbalance settings

And in the end in Results we can say: Ok, based on these different axises of analysis, then the results fall into these clusters (depending on what is missed from the report). 

So, this is not a replication study, but the study starts with replicating the base work. Because :

- First of all, we don’t want to imply something like these guys did not report on this, and this causes problems. We don’t want to target someone in specific and even it may be good if we could say something like: “we picked this paper because there was a reasonable amount of information available which allowed the reproducibility study, but even in these kinds of papers this range of results are possible”.

- And second, depends on who picks a study/paper, and depends on their experience/background, they can be like: from this chart you could understand this, or even though they didn't mention this, but that is irrelevant or this is obvious, etc. For example: for aggregation technique, may one may say the vote-based is irrelevant for this paper cause we have AUC diagrams. But sometimes there may be a case where we are looking for one prediction out of all classifiers (multi-class classification) which is the case for vote-based aggregation.

@glatard, @gkiar,  Please let me know what you think.

**Comment by gkiar** [comment]
Hi @HamidHeyde and @glatard -- this is now ready to review (and, I believe, merge) from my side. I replotted the data, regenerated the set of "best" models based on a somewhat more principled approach, and did a quantitative comparison of the model settings so that we could more concretely say which axes were important and which weren't.

There are two notable analyses which are missing from the results section, as they were incomplete before and I didn't want to overstep with my edits. @HamidHeyde, could you please do the following:

1. Re-run all of the models selected in the "best 10%" (easily findable from the new notebook in the repo) on multiple features
2. Re-run all of the models selected in the "best 10%" on the independent dataset

This work should not hold up merging this PR, but should be done separately and opened as another PR for @glatard and I to review.

Thanks!
Greg

**Comment by glatard** [comment]
Amazing work, thank you so much for doing that!
